### ISSUE TITLE
feature request: Support for EventBridge Pipes

### ISSUE THREAD

**Issue by drewko** [localstack/localstack#7920]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Feature description

AWS introduced EventBridge Pipes service on 1st December 2022. I am using it to integrate DynamoDb streams with SQS directly, and it would be nice to support it in localstack.

Right now aws cli commands return unimplemented error.
`aws pipes list-pipes --endpoint-url http://localhost:4566/`
`An error occurred (InternalFailure) when calling the ListPipes operation: API action 'ListPipes' for service 'pipes' not yet implemented or pro feature - check https://docs.localstack.cloud/user-guide/aws/feature-coverage for furthe
r information`

### 🧑‍💻 Implementation

https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes.html

**Comment by acostela** [comment]
Hi we're interested also on this. Are already any plan for this or can we help somehow? Thanks

**Comment by joe4dev** [comment]
> Hi we're interested also on this. Are already any plan for this or can we help somehow? Thanks

Hi @acostela and @drewko 

We are currently working on an initial version of EventBridge Pipes. What are the integrations (source, target, filter, enrichments) you are particularly interested?
* @drewko mentioned DynamoDB streams => SQS

**Comment by RobynVG** [comment]
Hi @joe4dev 

My team is also very interested in EventBridge pipes. I created the feature request in discuss https://discuss.localstack.cloud/t/event-bridge-pipes/428

For us we use a few different EventBridge pipes:

dynamo => SNS
dynamo => SQS

Each of them use a lambda for enrichment. We enrich and filter using the same lambdas so at least for us filter is low priority.

**Comment by endlessoblivion** [comment]
> We are currently working on an initial version of EventBridge Pipes. What are the integrations (source, target, filter, enrichments) you are particularly interested?
> 
>  - @drewko mentioned DynamoDB streams => SQS

The same here:
- DynamoDB Streams → SNS / SQS

**Comment by joe4dev** [comment]
Hi everyone 👋

**LocalStack Pro** now supports an initial version of EventBridge Pipes 🙌
* [LocalStack v3.1.0](https://github.com/localstack/localstack/releases/tag/v3.1.0) released initial Pipes support
* [LocalStack v3.2.0](https://github.com/localstack/localstack/releases/tag/v3.1.0) released further improvements


🟢 Our initial Pipes version implements:

- [API operations](https://docs.localstack.cloud/references/coverage/coverage_pipes/): `CreatePipe`, `DeletePipe`, `DescribePipe`
- CloudFormation resource `AWS::Pipes::Pipe`
- Sources: `Kinesis` (including SQS DLQ), `SQS` (including SQS DLQ), `DynamoDB`
- Basic filters
- Enrichments: `Lambda`
- Targets: `Kinesis`, `SQS`, `Lambda`, `SNS`
- Logging: `CloudWatch`
- Partial batch failure handling
- IAM enforcement for internal calls (see LocalStack [Security Testing](https://docs.localstack.cloud/user-guide/security-testing/))

🔴 The current version does not yet support:

- Input transformations
- Kinesis and DynamoDB resharding
- Concurrent pipe executions for high-throughput
- Validations and thorough exception handling (limited coverage and testing)

👉 Check out the user guide for more information: https://docs.localstack.cloud/user-guide/aws/pipes/

🙏 Thank you everyone for sharing your requirements and shaping LocalStack 🚀

**Comment by joe4dev** [comment]
❓Does our current implementation address your needs? How can we improve it?

**Comment by mustafakmal** [comment]
Hello,

Thanks for releasing support for Eventbridge pipes. In our use-case we want to configure a state machine as a target. Is this currently supported? I get the following error when trying to configure a state machine as a target for my eventbridge pipe

`An error occurred (InternalError) when calling the CreatePipe operation (reached max retries: 2): exception while calling pipes.CreatePipe: Unknown service: 'states'`

I am guessing that the statemachine arn depicted by 'states' is not yet supported?

Thanks!

**Comment by joe4dev** [comment]
Hi @mustafakmal 

Thank you for your feedback 🙏 

> I am guessing that the statemachine arn depicted by 'states' is not yet supported?

This is correct ATM, but we are happy to prioritize such feature requests upon feedback (we prioritize [support](https://docs.localstack.cloud/getting-started/help-and-support/) requests). Adding a StepFunctions target looks quite doable. What invocation type are you looking for in particular?
* Standard StepFunction: ASYNC invoke
* Express StepFunction: SYNC invoke or ASYNC invoke

**Comment by 70mmy** [comment]
Hello,

Any plans to support EventBridge in Targets?

I have an old system publishing SNS messages and a new system using EventBridge. Now it is configured like so SNS => SQS => EventBridge Pipe => EventBridge

Localstack looks so close to support this scenario as well, any thoughts if this can be supported?

**Comment by brookefrichards** [comment]
Chiming in here!

I am needing support for the Standard StepFunction as a target. In addition, i'm running this through terraform which calls ListTagsForResource, but i can make that feature request elsewhere if needed.

Thank you!

> This is correct ATM, but we are happy to prioritize such feature requests upon feedback (we prioritize [support](https://docs.localstack.cloud/getting-started/help-and-support/) requests). Adding a StepFunctions target looks quite doable. What invocation type are you looking for in particular?
> 
> * Standard StepFunction: ASYNC invoke
> * Express StepFunction: SYNC invoke or ASYNC invoke

**Comment by joe4dev** [comment]
> Hello,
> 
> Any plans to support EventBridge in Targets?
> 
> I have an old system publishing SNS messages and a new system using EventBridge. Now it is configured like so SNS => SQS => EventBridge Pipe => EventBridge
> 
> Localstack looks so close to support this scenario as well, any thoughts if this can be supported?

Thank you for your input @70mmy 🙏 It helps us prioritize 👍 
I added it to our backlog and will keep you updated here once implemented. It fits well because we are also improving our EventBridge support 📈

**Comment by joe4dev** [comment]
> > Hello,
> > Any plans to support EventBridge in Targets?
> > I have an old system publishing SNS messages and a new system using EventBridge. Now it is configured like so SNS => SQS => EventBridge Pipe => EventBridge
> > Localstack looks so close to support this scenario as well, any thoughts if this can be supported?
> 
> Thank you for your input @70mmy 🙏 It helps us prioritize 👍 I added it to our backlog and will keep you updated here once implemented. It fits well because we are also improving our EventBridge support 📈

@70mmy We just shipped support for the **EventBridge target** in the `latest` Docker image. The feature will also be part of the upcoming LocalStack v3.4 release scheduled for tomorrow 🚀 
We are looking forward to hearing your feedback 👂

**Comment by joe4dev** [comment]
> Chiming in here!
> 
> > This is correct ATM, but we are happy to prioritize such feature requests upon feedback (we prioritize [support](https://docs.localstack.cloud/getting-started/help-and-support/) requests). Adding a StepFunctions target looks quite doable. What invocation type are you looking for in particular?
> > 
> > * Standard StepFunction: ASYNC invoke
> > * Express StepFunction: SYNC invoke or ASYNC invoke
> 
> I am needing support for the Standard StepFunction as a target. In addition, i'm running this through terraform which calls ListTagsForResource, but i can make that feature request elsewhere if needed.
> 
> Thank you!

@brookefrichards @mustafakmal We just shipped support for the **Standard StepFunctions target** in the latest Docker image. The feature will also be part of the upcoming LocalStack v3.4 release scheduled for tomorrow 🚀
We are looking forward to hearing your feedback 👂

@brookefrichards `ListTagsForResource` is not yet supported

**Comment by 70mmy** [comment]
> > > Hello,
> > > Any plans to support EventBridge in Targets?
> > > I have an old system publishing SNS messages and a new system using EventBridge. Now it is configured like so SNS => SQS => EventBridge Pipe => EventBridge
> > > Localstack looks so close to support this scenario as well, any thoughts if this can be supported?
> > 
> > 
> > Thank you for your input @70mmy 🙏 It helps us prioritize 👍 I added it to our backlog and will keep you updated here once implemented. It fits well because we are also improving our EventBridge support 📈
> 
> @70mmy We just shipped support for the **EventBridge target** in the `latest` Docker image. The feature will also be part of the upcoming LocalStack v3.4 release scheduled for tomorrow 🚀 We are looking forward to hearing your feedback 👂

Hello @joe4dev

Thank you so much for the prompt implementation!I got around to play with it now, It looks like the configuration goes through fine, I can see the pipe created in the cli with the describe, that side is all good but, when I go to publish a message, it fails to push it to Event bridge. The message get's stuck in a loop, I get 5 times `AttributeError: 'str' object has no attribute 'get'` followed by a `Error while handling pipe events`

**Comment by 70mmy** [comment]
Hey @joe4dev 

I realised I can debug this myself in the docker container by playing around with the python files and I found the issue

I had this definition for the pipe in my serverless.yml

```    RecordStatePipe:
      Type: AWS::Pipes::Pipe
      Properties:
        Name: fabric-${sls:stage}-record-state-sqs-to-event-bus
        RoleArn: !GetAtt PipeRole.Arn
        Source: !GetAtt RecordStateSqs.Arn
        SourceParameters:
          SqsQueueParameters:
            BatchSize: 1
        TargetParameters:
          InputTemplate: >
            {
              "body": <$.body>
            }
        Target: !GetAtt EventBridge.Arn```

Notice the `InputTemplate`, in my case this was actually not needed and I could just removed it and it works as expected

What was actually happening, that `InputTemplate` would cause the source message to be passed forward as a string instead of a JSON / dict, so localstack would fail in `matches_event` with the error above, and of course later on, in my case the message was eventually consumed by a lambda, that lambda would get back a string as a `body` instead of the object it actually expected

I also have the `InputTemplate` deployed to AWS and there this does not happen, the `body` is passed around as an JSON / dict / object however you want to call it, so even if it is not required for me, I would say since it behaves differently it could be changed so localstack behaves the same as AWS for anyone that actually needs that input templace

**Comment by joe4dev** [comment]
@70mmy Thank you for sharing your feedback 🙏 

Transformations are indeed not yet implemented in EventBridge Pipes (see [limitations](https://docs.localstack.cloud/user-guide/aws/pipes/#limitations)).

Can you create a separate feature request to track this issue? It also helps to prioritize because others can comment and upvote.

**Comment by 70mmy** [comment]
Ah yes, my bad @joe4dev did not notice that limitation

I think the `InputTemplate` might be covered by this open issue https://github.com/localstack/localstack/issues/10287

**Comment by joe4dev** [comment]
> Ah yes, my bad @joe4dev did not notice that limitation
> 
> I think the `InputTemplate` might be covered by this open issue #10287

No worries. Thanks for raising it here anyway. It helps us to prioritize 🙏 

Issue #10287 is specific to EventBridge, but the implementation should be re-usable. I created an internal backlog item and will inform you once input transformers are supported for Pipes. In the meantime, I'm glad to know you have a workaround.

**Comment by joe4dev** [comment]
@drewko @acostela @RobynVG @endlessoblivion @70mmy @brookefrichards @mustafakmal Can we consider this feature request fulfilled with the current preview implementation of [EventBridge Pipes](https://docs.localstack.cloud/user-guide/aws/pipes/)?

EventBridge Pipes also benefited from recent behavioral improvements with Lambda Event Source Mappings (ESM) shipped with [LocalStack 3.8.0](https://blog.localstack.cloud/localstack-release-v-3-8-0/#new-default-lambda-event-source-mapping-implementation).

We are currently missing [API coverage](https://docs.localstack.cloud/references/coverage/coverage_pipes/) for `UpdatePipe`, `StartPipe`, and `StopPipe`.
The behavioral limitations are documented [here](https://docs.localstack.cloud/user-guide/aws/pipes/#current-limitations) including:
* Lack of input transformers.
* Lack of concurrency support (i.e., ParallelizationFactor), resulting in slower processing in high-throughput scenarios.
* Lack of lifecycle management for pipe states (i.e., missing tests for state transitions).
* Lack of re-sharding support when polling from Kinesis and DynamoDB streams.
* Batch handling behavior may have parity issues (e.g., batch flushing rules by size, length, time, etc. are not implemented).

Feel free to create separate GitHub issues for feature requests or any issue :)

**Comment by joe4dev** [comment]
> > Ah yes, my bad [@joe4dev](https://github.com/joe4dev) did not notice that limitation
> > I think the `InputTemplate` might be covered by this open issue [#10287](https://github.com/localstack/localstack/issues/10287)
> 
> No worries. Thanks for raising it here anyway. It helps us to prioritize 🙏
> 
> Issue [#10287](https://github.com/localstack/localstack/issues/10287) is specific to EventBridge, but the implementation should be re-usable. I created an internal backlog item and will inform you once input transformers are supported for Pipes. In the meantime, I'm glad to know you have a workaround.

Hi @70mmy 

Please follow this feature request regarding input transformers https://github.com/localstack/localstack/issues/12197 🚀
